app/models/picture.rb

```diff
@@ -20,29 +20,39 @@ class Picture
                          ['image/jpg', 'image/jpeg', 'image/png', 'image/gif'] }
 
   # SEE: http://stackoverflow.com/a/23141483/1064917
-  class << self
-    def from_url(file_location, parent)
-      pic = new(photographic: parent)
-      if Paperclip::Attachment.default_options[:storage].to_s != 'filesystem'
-        pic.attachment = open(file_location)
-      else # it's a filesystem update
-        # if it's already on the system, or it's missing,
-        # we don't need to update it.
-        unless file_location.include?('/system/') ||
-               file_location.include?('missing.png')
-          unless file_location.include?('http')
-            file_location = "#{Rails.root.join('public')}/#{file_location}"
-          end
-          pic.attachment = open(file_location)
-        end
-      end
-      pic.save!
-      pic
-    rescue StandardError => e
-      data = { file_location: file_location,
-               parent: parent }
-      ExceptionNotifier.notify_exception(e, data: data)
-      raise e
+  def self.from_url(file_location, parent)
+    pic = new(photographic: parent)
+    pic.do_attach(file_location)
+    pic.save!
+    pic
+  rescue StandardError => e
+    data = { file_location: file_location, parent: parent }
+    ExceptionNotifier.notify_exception(e, data: data)
+    raise e
+  end
+
+  def do_attach(file_location)
+    if Paperclip::Attachment.default_options[:storage].to_s != 'filesystem'
+      attach_via_network(file_location)
+    else
+      attach_via_filesystem(file_location)
     end
   end
+
+  def attach_via_network(file_location)
+    self.attachment = open(file_location)
```

Comment:
Can we avoid using direct `open` kernel calls,, especially on network locations.

Reply:
That's a good idea moving forward. I thought about replacing the calls with a more specific call, but that is beyond the scope of this PR ("Add tests for error handler in `Pictures.from_url()`"). This code was already in place and replacing the `Kernel#open` is not part of this PR. I can get to it in a future PR.

This is a pretty complex (and old) method. A refactor beyond moving stuff into smaller methods will require some manual QA, which was not my intent for this PR.